docs/DesignDocs/SGX_EMM.md

```diff
@@ -0,0 +1,678 @@
+ *                           The argument n specifies the binary logarithm of
+ *                           the desired alignment and must be at least 12.
+ *        Optionally ORed with one of following page types:
+ *             - SGX_EMA_PAGE_TYPE_REG: regular page type. This is the default if not specified.
+ *             - SGX_EMA_PAGE_TYPE_SS_FIRST: the first page in shadow stack.
+ *             - SGX_EMA_PAGE_TYPE_SS_REST: the rest page in shadow stack.
+ *
+ * @param[in] handler custom handler for page faults in this region, NULL if
+ *                     no custom handling needed.
+ * @param[in] handler_private private data for the @handler, which will be passed
+ *                     back when the handler is called.
+ * @param[out] out_addr pointer to store the start address of allocated range.
+ *                     Set to valid address by the function on success, NULL otherwise.
+ * @retval 0 The operation was successful.
+ * @retval EACCES Region is outside enclave address space.
+ * @retval EEXIST Any page in range requested is in use and SGX_EMA_FIXED is set.
+ * @retval EINVAL Invalid alignment bouandary, i.e., n < 12 in SGX_EMA_ALIGNED(n).
+ * @retval ENOMEM Out of memory, or no free space to satisfy alignment boundary.
+ */
+int sgx_mm_alloc(void *addr, size_t length, int flags,
+                 sgx_enclave_fault_handler_t handler, void *handler_private,
+                 void **out_addr);
+
+```
+
+**Remarks:**
+- Permissions of newly allocated regions are always SGX_EMA_PROT_READ|SGX_EMA_PROT_WRITE and of page
+    type SGX_EMA_PAGE_TYPE_REG, except for SGX_EMA_RESERVE mode regions which will have SGX_EMA_PROT_NONE.
+- Once allocated by sgx_mm_alloc, a region will stay in allocated state and become
+    deallocated once sgx_mm_dealloc is called.
+- If sgx_mm_dealloc on a partial range of a previously allocated region, then the
+    region is split, and the freed range is deallocated. The remainder of the
+    region stays allocated.
+- If all pages in the region are freed by sgx_mm_dealloc, then the whole region
+    is released, and the memory manager no longer tracks the region.
+
+
+### sgx_mm_uncommit and sgx_mm_dealloc
+
+```
+/*
+ * Uncommit (trim) physical EPC pages in a previously committed range.
+ * The pages in the allocation are freed, but the address range is still reserved.
+ * @param[in] addr page aligned start address of the region to be trimmed.
+ * @param[in] length size in bytes of multiples of page size.
+ * @retval 0 The operation was successful.
+ * @retval EINVAL The address range is not allocated or outside enclave.
+ */
+int sgx_mm_uncommit(void *addr, size_t length);
+
+/*
+ * Deallocate the address range.
+ * The pages in the allocation are freed and the address range is released for future allocation.
+ * @param[in] addr page aligned start address of the region to be freed and released
+ * @param[in] length size in bytes of multiples of page size.
```

Comment:
```suggestion
 * @param[in] length size in bytes in multiples of page size.
```

Reply:
Some places in this doc have "of multiples" and others say "in multiples".  I am not sure if they are intended to be the same.  Does this mean that the length must be a multiple of page size, as opposed to being the number of multiples of page size?

Reply:
length should be in bytes but the number value should be multiples of page size, k*4096. So is `of multiples of page size` better?

Reply:
```suggestion
 * @param[in] length size in bytes.  The length must be a multiple of the page size.
```